stl/inc/algorithm

```diff
@@ -132,12 +286,147 @@ _SourceTy* for_each_n(
 #endif // _ITERATOR_DEBUG_ARRAY_OVERLOADS
 #endif // _HAS_CXX17
 
+#ifdef __cpp_lib_concepts
+namespace ranges {
+    // ALIAS TEMPLATE for_each_result
+    template <class _It, class _Fn>
+    using for_each_result = in_fun_result<_It, _Fn>;
+
+    // VARIABLE ranges::for_each
+    class _For_each_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        template <input_iterator _It, sentinel_for<_It> _Se, class _Pj = identity,
+            indirectly_unary_invocable<projected<_It, _Pj>> _Fn>
+        constexpr for_each_result<_It, _Fn> operator()(_It _First, _Se _Last, _Fn _Func, _Pj _Proj = {}) const {

```

Comment:
I know this is hot under the wire for .6 but for future algorithms like this I think we need non-_Proj versions so that upgrading to the `ranges` version doesn't cause serious perf regressions.

Reply:
All three of these comments:
* non-projected overloads (or `if constexpr`ed "don't project when the projection is `identity`)
* reusing bodies with both `std` and `ranges` algorithms
* engaging low-level optimizations (`memcpy`, `memcmp`)
 are interrelated in interesting ways. After thinking through some of the alternatives, I decided they are out of scope for this PR: I don't want to rush out a design, and I *especially* don't want to break anything in `std` and delay the release.

Reply:
Note to self: collect these handwavy promises into the tentative Ranges plan-of-attack and post the result in #39.